core/meterer/meterer.go

```diff
@@ -138,9 +130,10 @@ func (m *Meterer) ValidateQuorum(headerQuorums []uint8, allowedQuorums []uint8)
 // ValidateBinIndex checks if the provided bin index is valid
 func (m *Meterer) ValidateBinIndex(header core.PaymentMetadata, reservation *core.ActiveReservation) bool {
 	now := uint64(time.Now().Unix())
-	currentBinIndex := GetBinIndex(now, m.ReservationWindow)
+	reservationWindow := m.ChainState.GetReservationWindow()
```

Comment:
does it make onchain query every time? should we cache these?

Reply:
I don't expect this to make onchain query every time. `ChainState` tracks a `PaymentVaultParams` struct that caches all the contract constants so this should query directly from the cache ([function](https://github.com/Layr-Labs/eigenda/blob/03b04753e5264a0cafb96bff676ec7651b36fb2c/core/meterer/onchain_state.go#L222)) and I expect onchain reads only when 1) [initializing](https://github.com/Layr-Labs/eigenda/blob/03b04753e5264a0cafb96bff676ec7651b36fb2c/core/meterer/onchain_state.go#L49) `ChainState struct` and 2) [refreshed](https://github.com/Layr-Labs/eigenda/blob/03b04753e5264a0cafb96bff676ec7651b36fb2c/core/meterer/onchain_state.go#L136) periodically along with registered payments

Reply:
Is there some kind of retry on failure mechanism?

Reply:
nada, although I don't think a regular read from cached struct will have any failure. There's periodic polling against the on-chain contract to refresh the state, so if one poll fails, the next poll will be the next try

Reply:
lol thought `ChainState` was this [ChainState](https://github.com/Layr-Labs/eigenda/blob/master/core/eth/state.go#L11). Should we just call this `OnchainPaymentState` or sth?

Reply:
yep we could do that! updated so it is more clear:)